### PR TITLE
fixed nthroot for prime powers

### PR DESCRIPTION
**Pull request by abh2k**
* ntheory
  * Fixes `nthroot_mod` for prime powers

### CODE REVIEW

File: sympy/ntheory/residue_ntheory.py [hunk 1]
```diff
@@ -760,9 +760,10 @@ def _nthroot_mod_composite(a, n, m):
                 diff = (rootn // (root or 1) * n) % p
                 if diff != 0:
                     ppow = p
+                    m_inv = mod_inverse(diff, p)
                     for j in range(1, e):
                         ppow *= p
-                        root = (root - (rootn - a) * mod_inverse(diff, p)) % ppow
+                        root = (root - (pow(root, n) - a) * m_inv) % ppow
```
Comment (abhinav28071999) [inline sympy/ntheory/residue_ntheory.py]: root changes on each iteration so rootn will also change. @jksuom

**Review by smichr — COMMENTED** [review]
Could you make the upper change using rootn, too? Just put pow(root, n-1) in the diff declaration. 

When I checked coverage and hand-tested some of your tests all looked ok. It looks like you have added some more tests. Does your code generated values of e that run through the e loop more than once so we know that the plow calc is right?

**Comment by smichr** [comment]
looks good

**Comment by abhinav28071999** [comment]
> Could you come the upper change using rootn, too? Just put pow(root, n-1) in the diff declaration.
> 
> When I checked coverage and hand-tested some of your tests all looked ok. It looks like you have added some more tests. Does your code generated values of e that run through the e loop more than once so we know that the plow calc is right?

In previous test cases while running through the e loop i noticed that the root were equal to 1, that's why we missed this mistake.
Just to be sure i am conducting some test right now on this function to see if there is any error.
Untitled

**Comment by smichr** [comment]
> i am conducting some test right now on this function to see if there is any error

```python
>>> from sympy import this
The Zen of SymPy
...
Coverage by testing matters.
Smart tests are better than random tests.
But random tests sometimes find what your smartest test missed.
...
```

**Comment by smichr** [comment]
`nthroot_mod(2,3,7,True)` returns None instead of `[]`.
```diff
diff --git a/sympy/ntheory/residue_ntheory.py b/sympy/ntheory/residue_ntheory.py
index 566b74d..aa0c31d 100644
--- a/sympy/ntheory/residue_ntheory.py
+++ b/sympy/ntheory/residue_ntheory.py
@@ -819,7 +819,7 @@ def nthroot_mod(a, n, p, all_roots=False):
     if a % p == 0:
         return [0]
     if not is_nthpow_residue(a, n, p):
-        return None
+        return [] if all_roots else None
     if (p - 1) % n == 0:
         return _nthroot_mod1(a, n, p, all_roots)
     # The roots of ``x**n - a = 0 (mod p)`` are roots of
```

**Pull request merged**